core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong128HashSet.java

```diff
@@ -0,0 +1,235 @@
+import io.questdb.std.Unsafe;
+import io.questdb.std.Vect;
+
+/**
+ * Specialized flyweight hash set used in {@link io.questdb.griffin.engine.functions.GroupByFunction}s.
+ * <p>
+ * Uses provided {@link GroupByAllocator} to allocate the underlying buffer. Grows the buffer when needed.
+ * <p>
+ * Buffer layout is the following:
+ * <pre>
+ * | capacity (in long128s) | size (in long128s) | size limit (in long128s) | padding | long128 array |
+ * +---------------------+-----------------+-----------------------+---------+------------+
+ * |       4 bytes       |     4 bytes     |       4 bytes         | 4 bytes |     -      |
+ * +---------------------+-----------------+-----------------------+---------+------------+
+ * </pre>
+ */
+public class GroupByLong128HashSet {
+    private static final long HEADER_SIZE = 4 * Integer.BYTES;
+    private static final int MIN_INITIAL_CAPACITY = 16;
+    private static final long SIZE_LIMIT_OFFSET = 2 * Integer.BYTES;
+    private static final long SIZE_OFFSET = Integer.BYTES;
+    private final int initialCapacity;
+    private final double loadFactor;
+    private final long noKeyValue;
+    private GroupByAllocator allocator;
+    private int mask;
+    private long ptr;
+
+    public GroupByLong128HashSet(int initialCapacity, double loadFactor, long noKeyValue) {
+        if (loadFactor <= 0d || loadFactor >= 1d) {
+            throw new IllegalArgumentException("0 < loadFactor < 1");
+        }
+        this.initialCapacity = Numbers.ceilPow2((int) (Math.max(initialCapacity, MIN_INITIAL_CAPACITY) / loadFactor));
+        this.loadFactor = loadFactor;
+        this.noKeyValue = noKeyValue;
+    }
+
+    /**
+     * Adds key to hash set preserving key uniqueness.
+     *
+     * @param lo low bits of key to be added.
+     * @param hi high bits of key to be added.
+     * @return false if key is already in the set and true otherwise.
+     */
+    public boolean add(long lo, long hi) {
+        int index = keyIndex(lo, hi);
+        if (index < 0) {
+            return false;
+        }
+        addAt(index, lo, hi);
+        return true;
+    }
+
+    public void addAt(int index, long lo, long hi) {
+        setKeyAt(index, lo, hi);
+        int size = size();
+        int sizeLimit = sizeLimit();
+        Unsafe.getUnsafe().putInt(ptr + SIZE_OFFSET, ++size);
+        if (size >= sizeLimit) {
+            rehash(capacity() << 1, sizeLimit << 1);
+        }
+    }
+
+    public int capacity() {
+        return ptr != 0 ? Unsafe.getUnsafe().getInt(ptr) : 0;
+    }
+    public long loKeyAt(int index) {
+        return Unsafe.getUnsafe().getLong(ptr + HEADER_SIZE + 16L * index);
+    }
+    public long hiKeyAt(int index) {
+        return Unsafe.getUnsafe().getLong(ptr + HEADER_SIZE + (16L * index) + 8L);
+    }
+
+    public int keyIndex(long lo, long hi) {
+        int hashCode = Hash.hashLong128(lo, hi);
+        int index = hashCode & mask;
+        long loKey = loKeyAt(index);
```

Comment:
It might be a good idea to change keyAt method to return the start address of the key in this and long256 set. Then accessing both parts of the key using `Unsafe` will only need trivial arithmetic operations.